### DISCUSSION TITLE
[Question][Hw4]

### DISCUSSION THREAD

**Discussion by anisrashidov** [prosyslab-classroom/cs348-information-security#709] [Q&A]
Can you describe what strings stand for in Open, Read_with_capability, Execute_with_capability commands?
type command =
.
.
.
| Open of string * Object.t * permission
| Execute_with_capability of string * Object.t * string
| Read_with_capability of string * Object.t

**Comment by sujin0529** [comment]
Hi,

Please refer to [comment](https://github.com/prosyslab-classroom/cs348-information-security/issues/281#issuecomment-1533938073).
If you still have questions, please let me know here again! :)

Thanks.

**Comment by anisrashidov** [comment]
Thank you. I read that. However, I still don't know exactly what it means.
In Read_with_capability command, I don't understand what "it seems that the first string is the name of capability (like 'fd' for example)." means. Can you give me an example to understand what the first string stands for?
Same applies to Execute_with_capability

**Comment by ahnjonggun** [comment]
I tried to print out the parameters of Execute_with_capability, which is string * object.t * string, in scenario 2.
Scenario2:
```
const enum access_control_t access_control = Capability;

void init() {
  create_subject("attacker");
  create_subject("deputy");
  create_object("secret_file");
  create_object("deputy");
  enter(Read, "attacker", "secret_file");
  enter(Execute, "attacker", "deputy");
}

void attacker() {
  int fd = open("secret_file", Read);
  execute_with_capability(fd, "deputy");
}

void deputy(int fd) { read_with_capability(fd, "secret_file"); }
```
My code in capability.run:
```
      | Execute_with_capability (str1, obj, str2) -> print_endline "execute with cap";
        print_endline str1;
        print_endline (Object.to_string obj);
        print_endline str2;
```
The result was 
> fd
> deputy
> fd

Since the document said that 
> For example, Execute_with_capability (perm, object_file, proc) will execute the target process proc with the capability perm to access object object_file

I can understand that first string being "fd" for capability, but why does the second string print out "fd" also?
I thought that it should be like "fd", "secret file" "deputy"

Thank you

**Comment by tanapthetimid** [comment]
I think the documentation is extremely misleading, and makes this homework problem confusing for no reason. for `Execute_with_capability (perm, object_file, proc)`, `perm` is the string name of the file descriptor argument, so in this case the `fd` in `execute_with_capability(fd, "deputy");`, and `proc` is the string name of the file descriptor parameter of the object, so the `fd` in `void deputy(int fd) {...`. And `object_file` is the thing being executed, so it becomes the subject as well.

**Comment by anisrashidov** [comment]
Why do we need the fd parameter string?

**Comment by sujin0529** [comment]
I think this [issue](https://github.com/prosyslab-classroom/cs348-information-security/issues/286#issue-1698879110) will answer your question.

You can think about why the file descriptor is used in capability to find the reason why the fd parameter is needed.
